### PR TITLE
[tests] Add a README.md to `Aspire.EndToEnd.Tests`

### PR DESCRIPTION
**Pull request by radical**
###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3061)

### CODE REVIEW

File: tests/Aspire.EndToEnd.Tests/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Aspire.EndToEnd.Tests
+
+## TL;dr or How do I use this?
+
+### Steps to prepare for `out-of-repo` runs
```
Comment (joperezr) [inline tests/Aspire.EndToEnd.Tests/README.md:5]: Correct me if I'm wrong, but all of these out-of-repo instructions are mainly for helix setup, right?

I assume most people checking this readme will be interested in learning about these tests, adding new ones, and running them locally (in-tree). Nothing wrong with the instructions you have, but might be good to add at the top a description of what the tests are for, briefly explain an overview of how they work (meaning how they launch testproject.apphost, etc) and how to add new tests. And then we can explain the out of tree and how to set it up.

Anyway, don't block on this feedback, but perhaps consider it for expanding this readme in the future.

Comment (joperezr) [reply tests/Aspire.EndToEnd.Tests/README.md:5]: Oh, I see you go over this a bit bellow, sorry for missing it I originally thought the contents were the same as your gist. Anyway, as mentioned above, feel free to ignore.

File: tests/Aspire.EndToEnd.Tests/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Aspire.EndToEnd.Tests
+
+## TL;dr or How do I use this?
```
Comment (RussKie) [inline tests/Aspire.EndToEnd.Tests/README.md:3]: [nit] Either all caps or the sentence case for "tl;dr"

Comment (radical) [inline tests/Aspire.EndToEnd.Tests/README.md:3]:
```suggestion
## TL;DR or How do I use this?
```

File: tests/Aspire.EndToEnd.Tests/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# Aspire.EndToEnd.Tests
+
+## TL;DR or How do I use this?
+
+### Steps to prepare for `out-of-repo` runs
+
+1. `.\build.cmd -pack` - to build all the nugets
+2. `dotnet build tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.csproj /t:InstallWorkloadUsingArtifacts /p:Configuration=<config>`
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/README.md:8]:
```suggestion
2. `dotnet build tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj /t:InstallWorkloadUsingArtifacts /p:Configuration=<config>`
```

File: tests/Aspire.EndToEnd.Tests/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    - this will install the sdk, and the `aspire` workload using the nugets from `artifacts/packages` into `artifacts/bin/dotnet-latest`
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/Aspire.EndToEnd.Tests/README.md:9]: Or clear some dotnet new template cache place?

File: tests/Aspire.EndToEnd.Tests/README.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    - note: `artifacts/bin/dotnet-none` contains the sdk+aspire workload manifest but NO workload
+
+- at this point the sdk is usable outside the repo too by add `artifacts/bin/dotnet-latest` to `PATH`. Note: you need to add `artifacts/packages/$(Configuration)/Shipping` as a nuget source for your projects, so the locally built packages can be picked up.
+
+### Using it from VS
+
+- No difference when running the EndToEnd tests in `in-repo` mode
+- For switching to `out-of-repo` add `<TestsRunningOutsideOfRepo>true</TestsRunningOutsideOfRepo>` to `tests/Aspire.EndToEnd.Tests/Directory.props` *before* any imports.
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/README.md:17]:
```suggestion
- For switching to `out-of-repo` add `<TestsRunningOutsideOfRepo>true</TestsRunningOutsideOfRepo>` to `tests/Aspire.EndToEnd.Tests/Directory.Build.props` *before* any imports.
```

File: tests/Aspire.EndToEnd.Tests/README.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    - tests cannot be run at this point as they will fail complaining about `artifacts/bin/dotnet-latest` being missing
+    - Install the sdk+workload following the steps above
+    - Run/debug the tests normally now, and they will be using the sdk
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/README.md:20]: I'm getting the following error from running the tests in the VS Test Explorer:

```
 Aspire.EndToEnd.Tests.IntegrationServicesTests.VerifyComponentWorks(resourceName: redis)
   Source: IntegrationServicesTests.cs line 34
   Duration: 1 ms

  Message: 
 Expected 0 exit code but got 1: C:\git\aspire2\artifacts\bin\dotnet-latest\dotnet.exe build -bl:C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testproject-build.binlog -v m 
Standard Output:
[build] MSBuild version 17.9.4+90725d08d for .NET
[build]   Determining projects to restore...
[build]   Restored C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.ServiceC\TestProject.ServiceC.csproj (in 89 ms).
[build]   Restored C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.ServiceA\TestProject.ServiceA.csproj (in 89 ms).
[build]   Restored C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.ServiceB\TestProject.ServiceB.csproj (in 89 ms).
[build]   Restored C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.WorkerA\TestProject.WorkerA.csproj (in 6.55 sec).
[build]   Restored C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.AppHost\TestProject.AppHost.csproj (in 14.03 sec).
[build]   Restored C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.IntegrationServiceA\TestProject.IntegrationServiceA.csproj (in 23.45 sec).
[build]   TestProject.ServiceC -> C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.ServiceC\bin\Debug\net8.0\TestProject.ServiceC.dll
[build]   TestProject.ServiceB -> C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.ServiceB\bin\Debug\net8.0\TestProject.ServiceB.dll
[build]   TestProject.ServiceA -> C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.ServiceA\bin\Debug\net8.0\TestProject.ServiceA.dll
[build]   TestProject.WorkerA -> C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.WorkerA\bin\Debug\net8.0\TestProject.WorkerA.dll
[build] CSC : error CS1705: Assembly 'Aspire.Npgsql' with identity 'Aspire.Npgsql, Version=42.42.42.42, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' uses 'Npgsql, Version=9.0.0.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7' which has a higher version than referenced assembly 'Npgsql' with identity 'Npgsql, Version=8.0.2.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7' [C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.IntegrationServiceA\TestProject.IntegrationServiceA.csproj]
[build] 
[build] Build FAILED.
[build] 
[build] CSC : error CS1705: Assembly 'Aspire.Npgsql' with identity 'Aspire.Npgsql, Version=42.42.42.42, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' uses 'Npgsql, Version=9.0.0.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7' which has a higher version than referenced assembly 'Npgsql' with identity 'Npgsql, Version=8.0.2.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7' [C:\git\aspire2\artifacts\bin\Aspire.EndToEnd.Tests\Debug\net8.0\testassets\testproject\TestProject.IntegrationServiceA\TestProject.IntegrationServiceA.csproj]
[build]     0 Warning(s)
[build]     1 Error(s)
[build] 
[build] Time Elapsed 00:00:32.93
[build] 
[build] Workload updates are available. Run `dotnet workload list` for more information.


  Stack Trace: 
CommandResult.EnsureExitCode(Int32 expectedExitCode, String messagePrefix, Boolean suppressOutput) line 44
CommandResult.EnsureSuccessful(String messagePrefix, Boolean suppressOutput) line 28
<<InitializeAsync>g__BuildProjectAsync|3>d.MoveNext() line 214
--- End of stack trace from previous location ---
IntegrationServicesFixture.InitializeAsync() line 73
```

Comment (eerhardt) [reply tests/Aspire.EndToEnd.Tests/README.md:20]: Same when running from `dotnet test` on the command line.

**Comment by danmoseley** [comment]
this looks obsolete now @radical ?

**Comment by radical** [comment]
> this looks obsolete now @radical ?

Yes. I was keeping it open to get to the problems that @eerhardt had. I have opened https://github.com/dotnet/aspire/issues/4784 to track that.
